### PR TITLE
add support for Lightimport in VRML (to DEV)

### PR DESCRIPTION
**Pull request by polygontwist**
for import PointLight and SpotLight. An AmbientLight is emitted when in the name "AmbientLight" occurs.
exampel: vrml-Code:

```
DEF Licht_AmbientLight Transform {
  translation 0 4.274257 0
  children [ 
    DEF LIGHT_Licht_AmbientLight PointLight {
      location 0 0 0
      radius 100000000
      color 0.34902 0.34902 1
      intensity 0.23
      on TRUE
    }
  ]
}

DEF Licht_1spot Transform {
  translation -56.342838 57.6591 -58.975361
  rotation -0.244504 0.769274 0.590284 -0.987861
  children [ 
    DEF LIGHT_Licht_1spot SpotLight {
      location 0 0 0
      direction 1 0 0
      cutOffAngle 0.587197
      radius 100000000
      color 0.47451 0.823529 1
      intensity 1
      on TRUE
    }
  ]
}

DEF Licht Transform {
  translation -49.414658 88.717239 51.611542
  children [ 
    DEF LIGHT_Licht PointLight {
      location 0 0 0
      radius 100000000
      color 1 0.807843 0.560784
      intensity 1
      on TRUE
    }
  ]
}
```

### CODE REVIEW

**Comment by mrdoob** [comment]
@bartmcleod looks good?

**Comment by bartmcleod** [comment]
@mrdoob @polygontwist The effort is much appreciated, of course, but we should not continue development on this type of VRML converter. There is a whole new package, that is inspired on this example in ThreeJs, that has more reliable VRML parsing, based on a PEG.js grammar. It uses ThreeJs for display. Light nodes can be easily added to it, so I would greatly appreciate if @polygontwist would direct his efforts there. New nodes are to be developed as separate classes in separate files, which makes for clearer code and easier debugging: https://github.com/bartmcleod/VrmlParser

This new package is not meant to compete with ThreeJs. It uses ThreeJs. It was developed becuase of support requests for the VrmlLoader, that could not be met by simply looping over the lines in the wrl file. It is being actively developed, so I invite @polygontwist to join.

**Comment by mrdoob** [comment]
@bartmcleod It's a shame that `VRMLLoader` can't sit with the other 30 dependency-less loaders... https://github.com/mrdoob/three.js/tree/dev/examples/js/loaders

I guess I'll have to give it another go at some point. I think there were VRML to X3D converters. Maybe a X3DLoader would be better and port the converter instead.

**Comment by mrdoob** [comment]
Until then, merging this!

**Comment by mrdoob** [comment]
Thanks @polygontwist!

**Comment by polygontwist** [comment]
@bartmcleod I had already written a parser until I had decided to change to tree.js. Thanks for the hint, I see in your project purely.

**Comment by bartmcleod** [comment]
@mrdoob VRML is just very hard to parse correctly. That's why I resorted to using a PEG.js grammar. It might be possible to just port the resulting js file, which can be used standalone. In doing so, it should be possible to integrate it.

**Comment by bartmcleod** [comment]
@mrdoob @polygontwist I managed to compile all of the code of the VrmlParser into a single 36kb js file, that could be included in three.js in lib. That way, the improved VrmlLoader could work without external dependencies. Would such an approach work for you?

**Comment by mrdoob** [comment]
@bartmcleod I think so!

**Comment by bartmcleod** [comment]
Looking at the directory structure of threejs and the way other loader examples are laid out, I am thinking of two different integrations:
1. I add only the vrml grammar parser in lib, as a dependency and use a `vmrl` directory in `examples/loaders` to put the ThreeJs renderer stuff in. With this setup, I would move the ThreeJs renderer code from my VrmlParser project, to the `examples/loaders/vrml` folder and keep only the grammar in my project. 
2. I compress everything related to VRML parsing and rendering in a single file in lib and change only the loader example html. 

The advantage of option 1 is that we can allow people to PR against the vrml rendering code from within the threejs project. The downside is that if they want to change the grammar itself, they will have to go to a different project.

The advantage of option 2 is that I can develop on both the grammar and the rendering code in the same project and the downside is that potential contributors will have to go to a project outside of three.js for both the rendering and the grammar code.

I feel inclined to aim for option 1, so that the grammar is in a separate project and the rendering just goes along with the other examples.

**Comment by mrdoob** [comment]
I would do 2. Following the sea3d approach.

**Comment by bartmcleod** [comment]
Thanks for the quick feedback! I will go with 2 then.

**Comment by bartmcleod** [comment]
Looking at what sea3d did, I would think option 1 is more like that, so now I'm confused.

The bottom line is:
If people want to contribute to rendering VRML nodes with Three.js, where should they submit their PRs, in the Three.js project (option 1) or in the VrmlParser project (option 2)?

**Comment by mrdoob** [comment]
Yeah, just realised that sea3d may not be a good example.

Have you looked at the X3D format? Maybe doing a X3DLoader and have a VRML2X3D.js converter as minified lib may also work.

**Comment by mrdoob** [comment]
By the way, the reason I'm suggesting this is because @tparisi himself suggested the same to me once.

**Comment by bartmcleod** [comment]
X3D is interesting, because it has nurbs ( if I recall correctly ), but I haven't worked with it ever. The reason I'm interested in the VRML format is that I used it almost twenty years ago and now I want to learn Three.js, by finding out how VRML translates into Three.js. The added benefit is that I can render old models in a browser without using a plugin. From what I know, the step from supporting VRML to supporting X3D as well should not be to big, but I have a lot of work ahead of me before I have full VRML support.

**Comment by bartmcleod** [comment]
Here is a PR that adds a better VRMLLoader, but still needs the light support. I'll work on that next, with the greatest priority, cherry picking the code from this issue.

https://github.com/mrdoob/three.js/pull/10290

**Comment by bartmcleod** [comment]
I have started building light support into VrmlParser. However, creating an ambient light based on the name of the node is not a good idea. This would only be reliable for those who have named their lights that way. It would however be possible to create an ambient light based on the accumulated ambient intensities of the lights in the scene.

**Comment by polygontwist** [comment]
you can delete this, i use cinema4d8 (it is the standartname) for export to webgl. i need ambientligth for a Projekt...... A little hack for more options.

**Comment by bartmcleod** [comment]
Ok, VrmlParser has now support for DirectionalLight, be it somewhat simplified: it can be improved upon. I will follow up with the Three.js integrated version. Those who want to use it only need to overwrite vrml.min.js in js/libs with the same file from VrmlParser.
